### DISCUSSION TITLE
How to use the "MULTI-EXPR" with meow-thing-register ?

### DISCUSSION THREAD

**Discussion by ghost** [meow-edit/meow#319] [Q&A] [answered]
Hi, I was trying to create new thing object and notice that function `meow-thing-register` support a "MULTI-EXPR", judging from the docstring it looks like a list whose item is  other EXPR form that `meow-thing-register` supports. 

My question is I can create a brackets thing that will match all kinds of brackets by using this "MULTI-EXPR"?

Edit:

 I tried out with something like this:
````elisp
(meow-thing-register 'brackets
                     '((pair ("(") (")"))
                       (pair ("[") ("]"))
                       (pair ("{") ("}")))
                     '((pair ("(") (")"))
                       (pair ("[") ("]"))
                       (pair ("{") ("}"))))
````
Which can match brackets, but it will always prioritize the round brackets over square ones, and again over curly ones... so for some thing like this:
````
( asdf [ asdfas { asdfkj |} asdf ] oiuoiu )
````
will match the outer most round brackets first, even if the cursor position is in the most deep nested curly brackets. 

What is the intended use of it anyway? And is there a preferred way to "merge" multiple thing into one?

**Comment by eshrh** [comment, marked as answer]
In a PAIR EXPR, each pair takes TOKENS, which are lists that can have multiple elements , like this:
```elisp
(meow-thing-register 'brackets
                     '(pair ("(" "[") (")" "]"))
                     '(pair ("(" "[") (")" "]")))
```
As far as i know, you can probably achieve the same thing by using a REGEXP-EXPR and using a bunch of conditionals. You might also be able to do it by using the syntax-table descriptions for opening delimiter and closing delimiter, which should be sanely defined per-mode.

However, this functionality is typically achieved by `meow-block`and `meow-block-expand`, is something missing from those commands that you'd like to see?

My issue with meow-block a while ago was that it does not let you immediately select the nesting level you are on, it instead snaps to the next deepest level, and you need to press it twice to get to your original level. Eventually though, I realized that most operations that involve doing specific operations at a specific nesting level can be achieved more smoothly via paredit, smartparens, and/or embrace.

**Reply by ghost** [reply]
Thank you, that dose the trick! 
While `meow-block` is sufficient for the most time, its behavior is depend on the major mode, in emacs-lisp-mode curly brackets won't recognized by `meow-block` for example, and I am trying to recreate the brackets text object that provides by the evil-text-obj package, that's why I want to "merge" different kinds of brackets thing into one.

> My issue with meow-block a while ago was that it does not let you immediately select the nesting level you are on, it instead snaps to the next deepest level,

Yeah I notice that too, I feels that moew.el don't want to be too "smart", meow tend to keeps a simple logic and avoid introducing ambiguity, that's why `(` `[` `{`  pairs are not merged into one thing I think.
